src/include/btree.i

```diff
@@ -492,23 +494,42 @@ static inline void
 __wt_page_only_modify_set(WT_SESSION_IMPL *session, WT_PAGE *page)
 {
 	uint64_t last_running;
+	uint8_t prev_mod_state;
+	bool made_dirty;
 
 	WT_ASSERT(session, !F_ISSET(session->dhandle, WT_DHANDLE_DEAD));
 
 	last_running = 0;
-	if (page->modify->write_gen == 0)
+	if (page->modify->mod_state == WT_MOD_STATE_CLEAN)
 		last_running = S2C(session)->txn_global.last_running;
 
 	/*
-	 * We depend on atomic-add being a write barrier, that is, a barrier to
+	 * We depend on atomic-cas being a write barrier, that is, a barrier to
 	 * ensure all changes to the page are flushed before updating the page
-	 * write generation and/or marking the tree dirty, otherwise checkpoints
+	 * mod state and/or marking the tree dirty, otherwise checkpoints
 	 * and/or page reconciliation might be looking at a clean page/tree.
 	 *
 	 * Every time the page transitions from clean to dirty, update the cache
 	 * and transactional information.
 	 */
-	if (__wt_atomic_add32(&page->modify->write_gen, 1) == 1) {
+	made_dirty = false;
+	prev_mod_state = page->modify->mod_state;
+	while (prev_mod_state < WT_MOD_STATE_MANY) {
+		if (__wt_atomic_cas8(&page->modify->mod_state,
+		    prev_mod_state, prev_mod_state + 1)) {
```

Comment:
We'd typically write this loop as `while ((prev_mod_state = page->modify->mod_state) < WT_MOD_STATE_MANY)` and save the separate reads before and at the end of the loop body.

That said, this code is already doing arithmetic on something we're trying to treat as an `enum` to calculate the new value, so why not use `__wt_atomic_add` (with a type larger than `uint8_t`)?

Reply:
Doesn't the `page->modify->mod_state < WT_MOD_STATE_MANY` race with the atomic increment and potentially result in a number greater than 2? The reason I'm doing the cas is to avoid this.

Reply:
It does potentially race: several threads could increment concurrently and the value could go over 2.  That's why I'm suggesting a type larger than `uint8_t`.  I don't think it changes the logic anywhere else, though: we only care if it's CLEAN, DIRTY or >=MANY.

Reply:
Understood. I've switched over to doing an increment now.